### ISSUE TITLE
Supabase CLI 1.123.0 started requiring docker to deploy edge functions

### ISSUE THREAD

**Issue by matart15** [supabase/supabase#19589]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

Supabase CLI 1.123.0 started requiring docker to deploy functions.

`pnpx supabase@1.123.0 functions deploy`
```
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?  
```

`pnpx supabase@1.112.0 functions deploy`
No problem

## To Reproduce

stop docker and run commands

## Expected behavior

Not require docker to deploy edge function

## System information

- OS: Windows + WSL

**Comment by ItzLue** [comment]
bump

**Comment by aLemonFox** [comment]
Something seems to have changed in the CLI that requires Docker to be running locally to deploy functions now. (1.123.4 latest as of writing this) Even when the function is only deployed to Supabase and the local development env is not used. When downgrading to 1.115.5 (most downloaded on npm) the issue is fixed.

**Comment by charislam** [comment]
This is a known change that we currently have documented: https://supabase.com/docs/guides/functions/deploy#deploy-your-edge-functions

**Comment by matart15** [comment]
@charislam 

Do you have discussion link for why team decided to require docker for edge function build?
I could not find it.

**Comment by charislam** [comment]
That's a good question! Let me page our Edge Functions team: @laktek

**Comment by andreasvirkus** [comment]
Is there any downside to deploying with version `1.112.0`?

**Comment by benjamindepoint** [comment]
This change has broken my Gitlab pipeline. All of my builds are on a docker container that has the Supabase CLI pre-installed. I cannot install a Docker desktop in a Docker image. What suggestion do you have?

**Comment by benjamindepoint** [comment]
> This change has broken my Gitlab pipeline. All of my builds are on a docker container that has the Supabase CLI pre-installed. I cannot install a Docker desktop in a Docker image. What suggestion do you have?

I solved this by switching to npx @ a lower version of the CLI versus installing the CLI with brew in the image. That's fine for now but I am pegged to a lower version which is not ideal. 

Please consider adding a flag to bypass the docker desktop check for such situations. Using a build image is a fairly common practice, and since the only way (that I know of) to deploy functions is via the CLI, this is limiting.

**Comment by encima** [comment]
[Since 1.123.4](https://supabase.com/docs/guides/functions/deploy#deploy-your-edge-functions), docker is a requirement for the CLI when deploying edge functions.

For Github, this should not be noticeable as runners default to the ubuntu vm where you can install docker.

For Gitlab, this is possible as well since the runners support [docker in docker (dind)](https://docs.gitlab.com/ee/ci/runners/saas/linux_saas_runner.html#docker-in-docker-support). Note that you will likely need to use the `docker-privileged` flag for this. See the linked docs for more info.

I will close this but please do comment with any further questions

**Comment by jmn** [comment]
If someone could please explain the rationale or need for requiring docker to deploy functions, especially since it still works to deploy them using an older version of supabase CLI, it would be much appreciated.

**Comment by sweatybridge** [comment]
Before version 1.123.4, the CLI deploys function by invoking a deno command to bundle the source files. The deno binary is downloaded from GitHub releases on demand. As you may expect, having binary dependencies this way is complicated and error prone. Especially since we support 3 different OS, 2 architectures, and the official macOS M1 build of deno doesn't exist. Using docker image means we only depend on arm64 and x86 builds of the edge runtime binary on Linux. Fewer points of failure means more stable cli and less maintenance work for us. 

The version of deno binary used for development is another issue. Our hosted edge runtime is updated independently by us, usually a few weeks after any official deno release. We had tickets in the past where users accidentally depend on new deno features for development but they are not available at runtime. Switching the bundling code to edge runtime means we can catch these problems at compile time before your code is deployed. 

And the last problem is with import map resolution for mono repos. While the old cli works for simple project structures, mono repos often import shared code that lives outside the supabase directory. When serving locally, we rewrite those import paths so they are accessible from inside the docker context. However, the rewrite doesn't happen for deployment so there could be inconsistencies between what is locally served and what is actually deployed. Using docker volume mount, we can eliminate such inconsistencies. 

As there are potentially other solutions to the problems above. We are happy to hear your suggestions on ways to improve the status quo.

**Comment by rienheuver** [comment]
My Gitlab deploys still fail because of this. The linked docs by @encima mention that `docker-privileged` is configured by default. Does that mean it should work by default? Because it doesn't for me. My job uses the `node:18` image. What can I do to make deploys work again?

**Comment by sweatybridge** [comment]
We have updated the ci/cd docs for deploying edge function on GitLab https://supabase.com/docs/guides/functions/cicd-workflow#gitlab-ci

Does it work for you?

**Comment by rienheuver** [comment]
Yes, thanks! Adding the following part got my job working again:
```
services:
    - docker:dind
  variables:
    DOCKER_HOST: tcp://docker:2375
```

**Comment by sfratini** [comment]
> We have updated the ci/cd docs for deploying edge function on GitLab https://supabase.com/docs/guides/functions/cicd-workflow#gitlab-ci
> 
> Does it work for you?

Could you please add examples and information throughout the documentation also with Azure DevOps pipelines? It is a decently used platform and it just does not work currently. I struggled for a bit to make this work. In any case, for people using macos agents in Azure DevOps, I had to make this to install Docker:

```
brew install colima docker
echo "Starting Colima..."
    colima start || colima start --cpu 2 --memory 4 --disk 10
    echo "Waiting for Docker to become available..."
    while ! docker info &>/dev/null; do
      echo "Waiting for Docker daemon..."
      sleep 5
    done
    echo "Docker is now running!"
```

The DockerInstall task in Azure only installs the cli but not the daemon. The "open -a Docker" does not work in command line so this is the only thing that worked for me. 

Unfortunately this makes the build run for a lot longer (over 10 minutes) and feels like a hack. I eventually realized that the Ubuntu images include Docker so I made a separated job in the release pipeline just for supabase, and this also works and makes the pipeline run in less than 2 minutes:

```
eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
docker version 
docker info
docker ps 
cd supabase
supabase functions deploy --project-ref $(project_id)
```

Hope this helps someone

**Comment by sweatybridge** [comment]
We have shipped an experimental flag for docker-less deploy to the beta release channel.

```
npx supabase@beta functions deploy --use-api
```

Let me know if you run into any issues using the above.

**Comment by benjamindepoint** [comment]
Thanks, will do!

- Benjamin De Point
"As Moses lifted up the serpent in the wilderness, even so must the Son of
Man be lifted up; so that whoever believes will in Him have eternal life."

**Comment by skyrover7** [comment]
> We have shipped an experimental flag for docker-less deploy to the beta release channel.
> 
> ```
> npx supabase@beta functions deploy --use-api
> ```
> 
> Let me know if you run into any issues using the above.

this worked! thank you
